File: submit/lab2/exercises/ast/ast.py
#!/usr/bin/env python3

import re
import sys
from collections import namedtuple
import json

"""
program
  : expr ';' program
  | #empty
  ;
expr
  : term ( ( '+' | '-' ) term )*
  ;
term
  : '-' term
  | factor
  ;
factor
  : INT
  | '(' expr ')'
  ;
"""

def parse(text):

    def check(kind): return lookahead.kind == kind
    def match(kind):
        nonlocal lookahead
        if (lookahead.kind == kind):
            lookahead = nextToken()
        else:
            print(f'expecting {kind} at {lookahead.lexeme}',
                  file=sys.stderr)
            sys.exit(1)
    def nextToken():
        nonlocal index
        if (index >= len(tokens)):
            return Token('EOF', '<EOF>')
        else:
            tok = tokens[index]
            index += 1
            return tok

    def program():
        asts = []
        while (not check('EOF')):
            asts.append(expr())
            match(';')
        return asts

    def expr():
        t = term()
        while (check('+') or (check('-'))):
            kind = lookahead.kind
            match(kind)
            t1 = term()
            t = Ast(kind, t, t1)
        return t

    def term():
        if (check('-')):
            match('-')
            return Ast('-', term())
        else:
            f = factor()
            if(check('**')):
                match('**')
                return Ast('**', f, term())
            return f

    def factor():
        if (check('INT')):
            value = int(lookahead.lexeme)
            match('INT')
            ast = Ast('INT')
            ast['value'] = value
            return ast
        else:
            match('(')
            value = expr()
            match(')')
            return value

    #begin parse()
    tokens = scan(text)
    index = 0
    lookahead = nextToken()
    value = program()
    if (not check('EOF')):
        print(f'expecting <EOF>, got {lookahead.lexeme}', file=sys.stderr)
        sys.exit(1)
    return value

def scan(text):
    def next_match(text):
        m = re.compile(r'\s+').match(text)
        if (m): return (m, None)
        m = re.compile(r'\d+').match(text)
        if (m): return (m, 'INT')
        m = re.compile(r'\*\*').match(text)
        if(m): return (m, '**')
        m = re.compile(r'.').match(text)  #must be last: match any char
        if (m): return (m, m.group())

    tokens = []
    while (len(text) > 0):
        (match, kind) = next_match(text)
        lexeme = match.group()
        if (kind): tokens.append(Token(kind, lexeme))
        text = text[len(lexeme):]
    return tokens

def main():
    if (len(sys.argv) != 2): usage();
    contents = readFile(sys.argv[1]);
    asts = parse(contents)
    print(json.dumps(asts, separators=(',', ':'))) #no whitespace

def readFile(path):
    with open(path, 'r') as file:
        content = file.read()
    return content


def usage():
    print(f'usage: {sys.argv[0]} DATA_FILE')
    sys.exit(1)

#use a dict so that we can add attributes dynamically
def Ast(tag, *kids):
    return { 'tag': tag, 'xkids': kids }

Token = namedtuple('Token', ['kind', 'lexeme'])

if __name__ == "__main__":
    main()

File: submit/lab2/exercises/ast/ast.mjs
#!/usr/bin/env node

import fs from 'fs';
import Path from 'path';


/* 
program
  : expr ';' program
  | #empty
  ;
expr
  : term ( ( '+' | '-' ) term )*
  ;
term
  : '-' term
  | factor
  ;
factor
  : INT
  | '(' expr ')'
  ;
*/

function parse(text) {
  const tokens = scan(text);
  let index = 0;
  let lookahead = nextToken();
  const value = program();
  return value;
  
  function check(kind) { return lookahead.kind === kind; }
  function match(kind) {
    if (check(kind)) {
      lookahead = nextToken();
    }
    else {
      console.error(`expecting ${kind} at ${lookahead.kind}`);
      process.exit();
    }
  }
  function nextToken() {
    return (
      (index >=  tokens.length) ? new Token('EOF', '<EOF>') : tokens[index++]
    );
  }

  function program() {
    const asts = [];
    while (!check('EOF')) {
      asts.push(expr());
      match(';');
    }
    return asts;
  }

  function expr() {
    let t = term();
    while (check('+') || check('-')) {
      const kind = lookahead.kind;
      match(kind);   
      const t1 = term();
      t = new Ast(kind, t, t1);
    }
    return t;
  }

  function term() {
    if (check('-')) {
      match('-');
      return new Ast('-', term());
    }
    else {
      return factor();
    }
  }

  function factor() {
    if (check('INT')) {
      const value = parseInt(lookahead.lexeme);
      match('INT');
      const ast = new Ast('INT');
      ast.value = value;
      return ast;
    }
    else {
      match('(');
      const value = expr();
      match(')');
      return value;
    }
  }
}


function scan(text) {
  const tokens = [];
  while (text.length > 0) {
    let m;
    if ((m = text.match(/^\s+/))) {
    }
    else if ((m = text.match(/^\d+/))) {
      tokens.push(new Token('INT', m[0]));
    }
    else {
      m = text.match(/^./);
      tokens.push(new Token(m[0], m[0]));
    }
    text = text.substring(m[0].length);
  }
  return tokens;
}


const CHAR_SET = 'utf8';
function main() {
  if (process.argv.length !== 3) usage();
  const file = process.argv[2];
  const text = fs.readFileSync(file, CHAR_SET);
  const value = parse(text);
  console.log(JSON.stringify(value));
}

function usage() {
  const prog = Path.basename(process.argv[1])
  console.error(`usage: ${prog} INPUT_FILE`);
  process.exit(1);
}

class Token {
  constructor(kind, lexeme) {
    Object.assign(this, {kind, lexeme});
  }
}

class Ast {
  constructor(tag, ...kids) {
    Object.assign(this, {tag, xkids: kids});
  }
}
main();

